Debugger: Add a Thread widget

sepalani opened the pull request:
~This a very early WIP PR.~ **This PR is ready to be reviewed.** I mostly would like feedback regarding the features that might be wanted for this widget and also the design since I'm really not a UI/UX person.

The goal of this new Widget is to display user thread information (not kernel/IOS thread). Regarding the pieces of information I can get:
 - Using the [Dolphin OS Globals](http://hitmen.c02.at/files/yagcd/yagcd/chap4.html#sec4.2.1)
    * current/defaut thread address
    * thread queue address
 - Using OSThread structure
    * thread's context (saved registers)
    * thread's stack
    * thread's state and attr
    * thread's priority
    * thread's error value
    * thread's data (often used by 3rd party libraries to store the thread name)
    * thread's callstack 
    * ...
 - Using LWPThread structure
    * Planned but out-of-scope of this PR 

Screenshot:
![image](https://user-images.githubusercontent.com/7890055/78633936-d36f7e80-78b3-11ea-9d72-ff89c94f37e0.png)

Addresses can be right-clicked to toggle a context menu:
![image](https://user-images.githubusercontent.com/7890055/79303159-45bd1000-7eff-11ea-98fe-2fe392282f38.png)

Any suggestions are welcome.

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,147 @@
+{
+  m_state = new QGroupBox(tr("State"));
+
+  auto create_line_edit = []() {
```
```suggestion
  const auto create_line_edit = [] {
```

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,147 @@
+  if (!isVisible() || Core::GetState() != Core::State::Paused)
+    return;
+
+  auto get_address_as_text = [](u32 addr) {
```
```suggestion
  const auto get_address_as_text = [](u32 addr) {
```

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.h:
```diff
@@ -0,0 +1,41 @@
+
+class QGroupBox;
+class QLineEdit;
+class QCloseEvent;
```
These should be alphabetically organized

lioncash commented on Source/Core/DolphinQt/Settings.cpp:
```diff
@@ -358,6 +358,21 @@ void Settings::SetRegistersVisible(bool enabled)
+
+void Settings::SetThreadsVisible(bool enabled)
+{
+  if (IsThreadsVisible() != enabled)
```
This can be inverted for an early return and to unindent the contained code.

sepalani commented:
Still WIP, the active threads group box was added.

![image](https://user-images.githubusercontent.com/7890055/78061962-0808a500-739f-11ea-979b-b5188a0461ad.png)

Data structures were reversed based on debug messages and assembly code from the GameCube ELF of Finding Nemo and the Wii ELF of Rayman Raving Rabbids.

The next steps will be to add the thread context group box and then add the context menu on table cells.

lioncash commented on Source/Core/Core/Debugger/Thread.cpp:
```diff
@@ -0,0 +1,159 @@
+//  - OSExceptionVector
+void Dolphin_Debugger::Thread::ReadContext(u32 addr, OSContext* ctx)
+{
+  for (int i = 0; i < 32; i += 1)
```
the array sizes can be queried from the arrays themselves, eliminating the need for magic numbers here.

It's also our convention to use `i++` or `++i` when incrementing by 1 within an iteration expression

lioncash commented on Source/Core/Core/Debugger/Thread.h:
```diff
@@ -0,0 +1,112 @@
+{
+  enum class State : u16
+  {
+    HAS_FPU = 1,
```
`HasFPU` and `HasException`

lioncash commented on Source/Core/Core/Debugger/Thread.h:
```diff
@@ -0,0 +1,112 @@
+};
+
+void ReadContext(u32 addr, OSContext* ctx);
+void DumpContext(OSContext* ctx);
```
Given this is dumping the context, this parameter can probably be const qualified. Ditto for `DumpMutex` and `DumpThread`

lioncash commented on Source/Core/Core/Debugger/Thread.h:
```diff
@@ -0,0 +1,112 @@
+void DumpMutex(OSMutex* mutex);
+void ReadThread(u32 addr, OSThread* thread);
+void DumpThread(OSThread* thread);
+bool IsThreadValid(OSThread* thread);
```
```suggestion
bool IsThreadValid(const OSThread* thread);
```

sepalani commented:
Compared to last time I added the thread's context and the thread's callstack:
![image](https://user-images.githubusercontent.com/7890055/78633936-d36f7e80-78b3-11ea-9d72-ff89c94f37e0.png)

Feature-wise I'm almost done. I need to do the dump functions and add dropdown menus on right-click.

I'm also planning to move the Threads functions to the PowerPC::debug_interface which IMO is more fitting. Then, I'll work on the UI/UX part and it should be ready to go.

sepalani commented:
This PR is ready to be reviewed.

Since last time:
 - I moved the Thread part to the DebugInterface
 - I added the context menu on right-click
 - I had to fix the WatchWidget to update the watch list on AddWatch

lioncash commented on Source/Core/Common/Debug/OSThread.cpp:
```diff
@@ -0,0 +1,205 @@
+//  - OSExceptionVector
+void Common::Debug::OSContext::Read(u32 addr, OSContext* ctx)
+{
+  for (int i = 0; i < 32; i++)
```
This loop and the ones below it can just query the array's size instead of hardcoding the value. `int` shouldn't be used here either since the functions expect an address, so this will cause sign mismatch warnings.

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,481 @@
+{
+  const auto get_hex = [](const std::optional<u32>& value) {
+    if (!value)
+      return QStringLiteral("");
```
```suggestion
      return QString{};
```

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,481 @@
+  };
+  const auto get_hex_idx = [](const auto& table, std::size_t index) {
+    if (!table || index >= table->size())
+      return QStringLiteral("");
```
```suggestion
      return QString{};
```

lioncash commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,481 @@
+  };
+  const auto get_f64_idx = [](const auto& table, std::size_t index) {
+    if (!table || index >= table->size())
+      return QStringLiteral("");
```
```suggestion
      return QString{};
```

sepalani commented:
@lioncash 
Done.

lioncash commented on Source/Core/Common/Debug/OSThread.h:
```diff
@@ -0,0 +1,151 @@
+  static void Read(u32 addr, OSContext* ctx);
+};
+
+static_assert(std::is_standard_layout_v<OSContext>);
```
Missed this in the previous review, but this is probably meant to be a `std::is_trivially_copyable_v`, not checking for standard layout.

sepalani replied:
IIRC, offsetof is UB if the struct/class isn't standard_layout. But you're also correct since I wanted to check if it's trivially copyable too.

sepalani replied:
@lioncash Done.

lioncash commented on Source/Core/Common/Debug/OSThread.h:
```diff
@@ -0,0 +1,151 @@
+  static bool IsValid(const OSThread* thread);
+};
+
+static_assert(std::is_standard_layout_v<OSThread>);
```
Ditto

lioncash commented on Source/Core/Common/Debug/OSThread.h:
```diff
@@ -0,0 +1,151 @@
+  OSThread m_thread;
+};
+
+static_assert(std::is_standard_layout_v<OSMutex>);
```
Ditto. These should probably be moved up underneath `OSMutex` as well, not under `OSThreadView`

sepalani commented:
Rebased on latest master, still ready to be reviewed & merged.

leoetlino commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,481 @@
+  UpdateThreadContext(&empty_context);
+}
+
+void ThreadWidget::UpdateThreadContext(const Common::Debug::PartialContext* context)
```
I'd pass a const-ref instead of a pointer here. Pointer semantics don't seem useful here.

And then you can just do `UpdateThreadContext({})` in the other functions to reset.

```suggestion
void ThreadWidget::UpdateThreadContext(const Common::Debug::PartialContext& context)
```

leoetlino commented on Source/Core/DolphinQt/Debugger/ThreadWidget.cpp:
```diff
@@ -0,0 +1,481 @@
+
+void ThreadWidget::UpdateThreadContext(const Common::Debug::PartialContext* context)
+{
+  const auto get_hex = [](const std::optional<u32>& value) {
```
same comment about get -> format

sepalani commented:
@leoetlino 
Done.

leoetlino commented on Source/Core/Common/Debug/OSThread.cpp:
```diff
@@ -0,0 +1,204 @@
+#include "Common/Debug/OSThread.h"
+
+#include <algorithm>
+#include "fmt/format.h"
```
Sorry for missing this last time, but this should be

```suggestion
#include <fmt/format.h>
```

leoetlino commented on Source/Core/Common/Debug/Threads.h:
```diff
@@ -0,0 +1,63 @@
+  virtual std::size_t GetStackSize() const = 0;
+  virtual s32 GetErrno() const = 0;
+  virtual std::string
+  GetSpecific() const = 0;  // Implementation specific, used to store arbitrary data
```
Could you put this comment on its own line, before `virtual std::string...`? Putting it here causes an awkward line break.

sepalani commented:
@leoetlino 
Done.
